tls.c

```diff
@@ -174,12 +176,10 @@ tls_new(int fd, bool client)
     return NULL;
   }
 
-  if (protocol != 0) {
-    errno = EINVAL; // FIXME
-    return NULL;
-  }
-
-  if (type != SOCK_STREAM && type != SOCK_DGRAM) {
+  if(!(
+    ((protocol == IPPROTO_TCP) && (type == SOCK_STREAM)) ||
+    ((protocol == IPPROTO_UDP) && (type == SOCK_DGRAM))) )
+  {
```

Comment:
I prefer separate if statements here. This makes it easy for a debugger to tell what is happening by getting the location of the return statement.